File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic/SingleLookupResult.cs
using System;
using System.Collections.Immutable;
using System.Linq;

namespace Microsoft.CodeAnalysis.VisualBasic
{
	internal struct SingleLookupResult
	{
		internal readonly LookupResultKind Kind;

		internal readonly Symbol Symbol;

		internal readonly DiagnosticInfo Diagnostic;

		public static readonly SingleLookupResult Empty = new SingleLookupResult(LookupResultKind.Empty, null, null);

		public static readonly SingleLookupResult EmptyAndStopLookup = new SingleLookupResult(LookupResultKind.EmptyAndStopLookup, null, null);

		public bool HasDiagnostic => Diagnostic != null;

		public bool StopFurtherLookup => Kind >= LookupResultKind.WrongArityAndStopLookup;

		public bool IsGoodOrAmbiguous
		{
			get
			{
				if (Kind != LookupResultKind.Good)
				{
					return Kind == LookupResultKind.Ambiguous;
				}
				return true;
			}
		}

		public bool IsGood => Kind == LookupResultKind.Good;

		public bool IsAmbiguous => Kind == LookupResultKind.Ambiguous;

		internal SingleLookupResult(LookupResultKind kind, Symbol symbol, DiagnosticInfo diagInfo)
		{
			this = default(SingleLookupResult);
			Kind = kind;
			Symbol = symbol;
			Diagnostic = diagInfo;
		}

		public static SingleLookupResult Good(Symbol sym)
		{
			return new SingleLookupResult(LookupResultKind.Good, sym, null);
		}

		public static SingleLookupResult Ambiguous(ImmutableArray<Symbol> syms, Func<ImmutableArray<Symbol>, AmbiguousSymbolDiagnostic> generateAmbiguityDiagnostic)
		{
			DiagnosticInfo diagInfo = generateAmbiguityDiagnostic(syms);
			return new SingleLookupResult(LookupResultKind.Ambiguous, syms.First(), diagInfo);
		}

		public static SingleLookupResult WrongArityAndStopLookup(Symbol sym, ERRID err)
		{
			return new SingleLookupResult(LookupResultKind.WrongArityAndStopLookup, sym, new BadSymbolDiagnostic(sym, err));
		}

		public static SingleLookupResult WrongArityAndStopLookup(Symbol sym, DiagnosticInfo diagInfo)
		{
			return new SingleLookupResult(LookupResultKind.WrongArityAndStopLookup, sym, diagInfo);
		}

		public static SingleLookupResult WrongArity(Symbol sym, DiagnosticInfo diagInfo)
		{
			return new SingleLookupResult(LookupResultKind.WrongArity, sym, diagInfo);
		}

		public static SingleLookupResult WrongArity(Symbol sym, ERRID err)
		{
			return new SingleLookupResult(LookupResultKind.WrongArity, sym, new BadSymbolDiagnostic(sym, err));
		}

		public static SingleLookupResult MustNotBeInstance(Symbol sym, ERRID err)
		{
			return new SingleLookupResult(LookupResultKind.MustNotBeInstance, sym, new BadSymbolDiagnostic(sym, err));
		}

		public static SingleLookupResult MustBeInstance(Symbol sym)
		{
			return new SingleLookupResult(LookupResultKind.MustBeInstance, sym, null);
		}

		public static SingleLookupResult Inaccessible(Symbol sym, DiagnosticInfo diagInfo)
		{
			return new SingleLookupResult(LookupResultKind.Inaccessible, sym, diagInfo);
		}

		internal static SingleLookupResult NotAnAttributeType(Symbol sym, DiagnosticInfo error)
		{
			return new SingleLookupResult(LookupResultKind.NotAnAttributeType, sym, error);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/SingleLookupResult.vb
Imports Microsoft.CodeAnalysis
Imports System
Imports System.Collections.Immutable
Imports System.Linq

Namespace Microsoft.CodeAnalysis.VisualBasic
	Friend Structure SingleLookupResult
		Friend ReadOnly Kind As LookupResultKind

		Friend ReadOnly Symbol As Microsoft.CodeAnalysis.VisualBasic.Symbol

		Friend ReadOnly Diagnostic As DiagnosticInfo

		Public ReadOnly Shared Empty As SingleLookupResult

		Public ReadOnly Shared EmptyAndStopLookup As SingleLookupResult

		Public ReadOnly Property HasDiagnostic As Boolean
			Get
				Return Me.Diagnostic IsNot Nothing
			End Get
		End Property

		Public ReadOnly Property IsAmbiguous As Boolean
			Get
				Return Me.Kind = LookupResultKind.Ambiguous
			End Get
		End Property

		Public ReadOnly Property IsGood As Boolean
			Get
				Return Me.Kind = LookupResultKind.Good
			End Get
		End Property

		Public ReadOnly Property IsGoodOrAmbiguous As Boolean
			Get
				If (Me.Kind = LookupResultKind.Good) Then
					Return True
				End If
				Return Me.Kind = LookupResultKind.Ambiguous
			End Get
		End Property

		Public ReadOnly Property StopFurtherLookup As Boolean
			Get
				Return Me.Kind >= LookupResultKind.WrongArityAndStopLookup
			End Get
		End Property

		Shared Sub New()
			SingleLookupResult.Empty = New SingleLookupResult(LookupResultKind.Empty, Nothing, Nothing)
			SingleLookupResult.EmptyAndStopLookup = New SingleLookupResult(LookupResultKind.EmptyAndStopLookup, Nothing, Nothing)
		End Sub

		Friend Sub New(ByVal kind As LookupResultKind, ByVal symbol As Microsoft.CodeAnalysis.VisualBasic.Symbol, ByVal diagInfo As DiagnosticInfo)
			Me = New SingleLookupResult() With
			{
				.Kind = kind,
				.Symbol = symbol,
				.Diagnostic = diagInfo
			}
		End Sub

		Public Shared Function Ambiguous(ByVal syms As ImmutableArray(Of Microsoft.CodeAnalysis.VisualBasic.Symbol), ByVal generateAmbiguityDiagnostic As Func(Of ImmutableArray(Of Microsoft.CodeAnalysis.VisualBasic.Symbol), AmbiguousSymbolDiagnostic)) As SingleLookupResult
			Dim diagnosticInfo As Microsoft.CodeAnalysis.DiagnosticInfo = generateAmbiguityDiagnostic(syms)
			Return New SingleLookupResult(LookupResultKind.Ambiguous, syms.First(), diagnosticInfo)
		End Function

		Public Shared Function Good(ByVal sym As Microsoft.CodeAnalysis.VisualBasic.Symbol) As SingleLookupResult
			Return New SingleLookupResult(LookupResultKind.Good, sym, Nothing)
		End Function

		Public Shared Function Inaccessible(ByVal sym As Microsoft.CodeAnalysis.VisualBasic.Symbol, ByVal diagInfo As DiagnosticInfo) As SingleLookupResult
			Return New SingleLookupResult(LookupResultKind.Inaccessible, sym, diagInfo)
		End Function

		Public Shared Function MustBeInstance(ByVal sym As Microsoft.CodeAnalysis.VisualBasic.Symbol) As SingleLookupResult
			Return New SingleLookupResult(LookupResultKind.MustBeInstance, sym, Nothing)
		End Function

		Public Shared Function MustNotBeInstance(ByVal sym As Microsoft.CodeAnalysis.VisualBasic.Symbol, ByVal err As ERRID) As SingleLookupResult
			Return New SingleLookupResult(LookupResultKind.MustNotBeInstance, sym, New BadSymbolDiagnostic(sym, err))
		End Function

		Friend Shared Function NotAnAttributeType(ByVal sym As Microsoft.CodeAnalysis.VisualBasic.Symbol, ByVal [error] As DiagnosticInfo) As SingleLookupResult
			Return New SingleLookupResult(LookupResultKind.NotAnAttributeType, sym, [error])
		End Function

		Public Shared Function WrongArity(ByVal sym As Microsoft.CodeAnalysis.VisualBasic.Symbol, ByVal diagInfo As DiagnosticInfo) As SingleLookupResult
			Return New SingleLookupResult(LookupResultKind.WrongArity, sym, diagInfo)
		End Function

		Public Shared Function WrongArity(ByVal sym As Microsoft.CodeAnalysis.VisualBasic.Symbol, ByVal err As ERRID) As SingleLookupResult
			Return New SingleLookupResult(LookupResultKind.WrongArity, sym, New BadSymbolDiagnostic(sym, err))
		End Function

		Public Shared Function WrongArityAndStopLookup(ByVal sym As Microsoft.CodeAnalysis.VisualBasic.Symbol, ByVal err As ERRID) As SingleLookupResult
			Return New SingleLookupResult(LookupResultKind.WrongArityAndStopLookup, sym, New BadSymbolDiagnostic(sym, err))
		End Function

		Public Shared Function WrongArityAndStopLookup(ByVal sym As Microsoft.CodeAnalysis.VisualBasic.Symbol, ByVal diagInfo As DiagnosticInfo) As SingleLookupResult
			Return New SingleLookupResult(LookupResultKind.WrongArityAndStopLookup, sym, diagInfo)
		End Function
	End Structure
End Namespace